// @bcoe/v8-coverage/dist/lib/index.js
"use strict";

function emitForest(cov, options) {
  // function body
}

function emitForestLines(cov, options) {
  // function body
}

function parseFunctionRanges(src, ranges) {
  // function body
}

function parseOffsets(src, offsets) {
  // function body
}

module.exports = {
  emitForest,
  emitForestLines,
  parseFunctionRanges,
  parseOffsets,
};


"use strict";

function cloneFunctionCov(fcov) {
  // function body
}

function cloneProcessCov(pcov) {
  // function body
}

function cloneScriptCov(scov) {
  // function body
}

function cloneRangeCov(rcov) {
  // function body
}

module.exports = {
  cloneFunctionCov,
  cloneProcessCov,
  cloneScriptCov,
  cloneRangeCov,
};


"use strict";

function compareScriptCovs(scov1, scov2) {
  // function body
}

function compareFunctionCovs(fcov1, fcov2) {
  // function body
}

function compareRangeCovs(rcov1, rcov2) {
  // function body
}

module.exports = {
  compareScriptCovs,
  compareFunctionCovs,
  compareRangeCovs,
};


"use strict";

function mergeFunctionCovs(covs) {
  // function body
}

function mergeProcessCovs(covs) {
  // function body
}

function mergeScriptCovs(covs) {
  // function body
}

module.exports = {
  mergeFunctionCovs,
  mergeProcessCovs,
  mergeScriptCovs,
};


"use strict";

class RangeTree {
  constructor(ranges) {
    // class body
  }

  // other class methods
}

module.exports = RangeTree;


"use strict";

const ascii = require('./ascii');
const clone = require('./clone');
const compare = require('./compare');
const merge = require('./merge');
const RangeTree = require('./range-tree');

module.exports = {
  ascii,
  clone,
  compare,
  merge,
  RangeTree,
};


"use strict";

const covUtils = require('./index');

// use the functions from the covUtils object
